File: src/frontend/src/services/instances.ts
import { apiDelete, apiGet, apiPost } from './apiClient';
import type { ApiResult } from './api.types';

export type PolarisInstanceKind = 'project' | 'bench_project' | 'internal_test' | string;

export interface PolarisInstance {
  schema_version: number;
  instance_id: string;
  name: string;
  kind: PolarisInstanceKind;
  polaris_root: string;
  workspace: string;
  runtime_root: string;
  backend_port: number;
  frontend_port: number;
  backend_url: string;
  frontend_url: string;
  token: string;
  backend_reload: boolean;
  frontend_vite: boolean;
  start_frontend: boolean;
  status: string;
  backend_pid: number | null;
  frontend_pid: number | null;
  backend_alive: boolean;
  frontend_alive: boolean;
  created_at: string;
  updated_at: string;
  last_started_at: string;
  last_stopped_at: string;
  bench: Record<string, unknown>;
  metadata: Record<string, unknown>;
}

export interface InstanceListResponse {
  instances: PolarisInstance[];
}

export interface InstanceResponse {
  instance: PolarisInstance;
}

export interface InstanceLogsResponse {
  stream: 'backend' | 'frontend';
  content: string;
}

export interface StartInstancePayload {
  instance_id?: string;
  name?: string;
  kind?: PolarisInstanceKind;
  polaris_root?: string;
  workspace: string;
  runtime_root?: string;
  backend_port?: number | null;
  frontend_port?: number | null;
  token?: string;
  backend_reload?: boolean;
  frontend_vite?: boolean;
  start_frontend?: boolean;
  bench?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
}

export async function listInstances(): Promise<ApiResult<InstanceListResponse>> {
  return apiGet<InstanceListResponse>('/v2/instances', '读取 Polaris 实例列表失败');
}

export async function startInstance(payload: StartInstancePayload): Promise<ApiResult<InstanceResponse>> {
  return apiPost<InstanceResponse>('/v2/instances/start', payload, '启动 Polaris 实例失败');
}

export async function stopInstance(instanceId: string): Promise<ApiResult<InstanceResponse>> {
  return apiPost<InstanceResponse>(
    `/v2/instances/${encodeURIComponent(instanceId)}/stop`,
    {},
    '停止 Polaris 实例失败',
  );
}

export async function restartInstance(instanceId: string): Promise<ApiResult<InstanceResponse>> {
  return apiPost<InstanceResponse>(
    `/v2/instances/${encodeURIComponent(instanceId)}/restart`,
    {},
    '重启 Polaris 实例失败',
  );
}

export async function deleteInstance(instanceId: string): Promise<ApiResult<{ ok: boolean }>> {
  return apiDelete<{ ok: boolean }>(
    `/v2/instances/${encodeURIComponent(instanceId)}`,
    '删除 Polaris 实例失败',
  );
}

export async function getInstanceLogs(
  instanceId: string,
  stream: 'backend' | 'frontend',
  tailLines = 300,
): Promise<ApiResult<InstanceLogsResponse>> {
  const params = new URLSearchParams({ stream, tail_lines: String(tailLines) });
  return apiGet<InstanceLogsResponse>(
    `/v2/instances/${encodeURIComponent(instanceId)}/logs?${params.toString()}`,
    '读取 Polaris 实例日志失败',
  );
}

export function buildInstanceWorkspaceUrl(instance: PolarisInstance): string {
  const base = instance.frontend_url || window.location.origin;
  const url = new URL(base);
  url.searchParams.set('instance', instance.instance_id);
  url.searchParams.set('backend', instance.backend_url);
  if (instance.workspace) {
    url.searchParams.set('workspace', instance.workspace);
  }
  if (instance.token) {
    url.searchParams.set('token', instance.token);
  }
  return url.toString();
}

File: src/backend/polaris/delivery/http/v2/instances.py
"""Platform instance management endpoints."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query
from polaris.cells.instances.public.service import get_instance_supervisor
from polaris.delivery.http.dependencies import require_auth
from pydantic import BaseModel, Field

router = APIRouter(prefix="/instances", tags=["Instances"])


class StartInstanceRequest(BaseModel):
    instance_id: str = ""
    name: str = ""
    kind: str = "project"
    polaris_root: str = ""
    workspace: str
    runtime_root: str = ""
    backend_port: int | None = None
    frontend_port: int | None = None
    token: str = ""
    backend_reload: bool = True
    frontend_vite: bool = True
    start_frontend: bool = True
    bench: dict[str, Any] = Field(default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)


@router.get("", dependencies=[Depends(require_auth)])
async def list_instances() -> dict[str, Any]:
    supervisor = get_instance_supervisor()
    return {"instances": supervisor.list_instances()}


@router.post("/start", dependencies=[Depends(require_auth)])
async def start_instance(request: StartInstanceRequest) -> dict[str, Any]:
    supervisor = get_instance_supervisor()
    try:
        instance = supervisor.start_instance(request.model_dump())
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except RuntimeError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from exc
    return {"instance": instance}


@router.post("/{instance_id}/stop", dependencies=[Depends(require_auth)])
async def stop_instance(instance_id: str) -> dict[str, Any]:
    supervisor = get_instance_supervisor()
    try:
        return {"instance": supervisor.stop_instance(instance_id)}
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except RuntimeError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc


@router.post("/{instance_id}/restart", dependencies=[Depends(require_auth)])
async def restart_instance(instance_id: str) -> dict[str, Any]:
    supervisor = get_instance_supervisor()
    try:
        return {"instance": supervisor.restart_instance(instance_id)}
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except RuntimeError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.delete("/{instance_id}", dependencies=[Depends(require_auth)])
async def delete_instance(instance_id: str) -> dict[str, Any]:
    try:
        deleted = get_instance_supervisor().delete_instance(instance_id)
    except RuntimeError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    if not deleted:
        raise HTTPException(status_code=404, detail="instance not found")
    return {"ok": True}


@router.get("/{instance_id}/health", dependencies=[Depends(require_auth)])
async def instance_health(instance_id: str) -> dict[str, Any]:
    supervisor = get_instance_supervisor()
    try:
        return {"instance": supervisor.health(instance_id)}
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.get("/{instance_id}/logs", dependencies=[Depends(require_auth)])
async def instance_logs(
    instance_id: str,
    stream: str = Query("backend", pattern="^(backend|frontend)$"),
    tail_lines: int = Query(200, ge=1, le=5000),
) -> dict[str, str]:
    supervisor = get_instance_supervisor()
    try:
        return {"stream": stream, "content": supervisor.get_logs(instance_id, stream, tail_lines)}
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
